File: apple/Shared/Models/CareConfidence.swift
import Foundation

enum ConnectionStatus: Sendable {
    case connecting
    case live
    case unavailable
}

enum CareConfidenceLevel: String, Sendable {
    case excellent = "Excellent"
    case good = "Good"
    case needsAttention = "Needs Attention"
    case critical = "Critical"
}

struct ConfidenceSignal: Identifiable, Sendable {
    let id = UUID()
    let name: String
    let value: String
    let isHealthy: Bool
}

struct CareConfidence: Sendable {
    let score: Int
    let level: CareConfidenceLevel
    let summary: String
    let signals: [ConfidenceSignal]

    static func evaluate(
        ping: LocationPing?,
        connection: ConnectionStatus,
        now: Date = .now
    ) -> CareConfidence {
        guard let ping else {
            return CareConfidence(
                score: 0,
                level: .needsAttention,
                summary: "Waiting for the patient device to share its location.",
                signals: [
                    ConfidenceSignal(name: "Location", value: "Not available", isHealthy: false),
                    ConfidenceSignal(name: "Connection", value: "Waiting", isHealthy: false)
                ]
            )
        }

        let age = max(0, now.timeIntervalSince(ping.timestamp))
        let accuracy = ping.accuracy ?? 100
        var score = 100

        if connection != .live { score -= 35 }
        if age > 120 { score -= 40 }
        else if age > 60 { score -= 25 }
        else if age > 20 { score -= 10 }

        if accuracy > 80 { score -= 25 }
        else if accuracy > 35 { score -= 15 }
        else if accuracy > 15 { score -= 7 }

        if let battery = ping.battery {
            if battery < 10 { score -= 25 }
            else if battery < 25 { score -= 10 }
        }

        switch ping.stateAtTime {
        case .alert: score -= 35
        case .grace: score -= 20
        case .caution: score -= 8
        case .safe, .unknown: break
        }

        score = min(100, max(0, score))
        let level: CareConfidenceLevel =
            score >= 90 ? .excellent :
            score >= 70 ? .good :
            score >= 40 ? .needsAttention : .critical

        let summary: String = switch level {
        case .excellent: "Location is current and the patient device is reporting clearly."
        case .good: "SafeZone has enough current information to provide a reliable status."
        case .needsAttention: "Some safety signals need attention."
        case .critical: "SafeZone cannot provide a reliable current status."
        }

        return CareConfidence(
            score: score,
            level: level,
            summary: summary,
            signals: [
                ConfidenceSignal(
                    name: "Last update",
                    value: ping.timestamp.formatted(.relative(presentation: .named)),
                    isHealthy: age <= 60
                ),
                ConfidenceSignal(
                    name: "Location clarity",
                    value: ping.accuracy.map { "About \(Int($0.rounded())) m" } ?? "Unknown",
                    isHealthy: accuracy <= 35
                ),
                ConfidenceSignal(
                    name: "Connection",
                    value: connection == .live ? "Live" : "Interrupted",
                    isHealthy: connection == .live
                ),
                ConfidenceSignal(
                    name: "Patient device",
                    value: ping.battery.map { "\(Int($0.rounded()))% battery" } ?? "Battery unavailable",
                    isHealthy: ping.battery.map { $0 >= 25 } ?? true
                ),
                ConfidenceSignal(
                    name: "Movement status",
                    value: ping.stateAtTime.title,
                    isHealthy: ping.stateAtTime == .safe
                )
            ]
        )
    }
}

File: src/careConfidence.ts
import type { GeofenceState, LocationPing } from "./types";
import { STATE_CHIP_LABEL } from "./safetyLanguage";

export type ConnectionState = "connecting" | "live" | "offline";
export type CareConfidenceLevel = "excellent" | "good" | "attention" | "critical";

export type CareConfidence = {
  score: number;
  level: CareConfidenceLevel;
  label: string;
  summary: string;
  details: Array<{
    label: string;
    value: string;
    healthy: boolean;
  }>;
};

function clamp(value: number) {
  return Math.max(0, Math.min(100, Math.round(value)));
}

function relativeUpdate(timestamp: string | null) {
  if (!timestamp) {
    return "No location yet";
  }

  const seconds = Math.max(0, Math.round((Date.now() - Date.parse(timestamp)) / 1000));
  if (seconds < 10) return "Just now";
  if (seconds < 60) return `${seconds} sec ago`;

  const minutes = Math.round(seconds / 60);
  return `${minutes} min ago`;
}

export function calculateCareConfidence(
  ping: LocationPing | null,
  connection: ConnectionState,
  state: GeofenceState,
  now = Date.now()
): CareConfidence {
  if (!ping) {
    return {
      score: 0,
      level: "attention",
      label: "Location unavailable",
      summary: "Open SafeZone on the patient phone and allow location access.",
      details: [
        { label: "Location", value: "Not available", healthy: false },
        { label: "Connection", value: connection === "live" ? "Connected" : "Not connected", healthy: false }
      ]
    };
  }

  const ageSeconds = Math.max(0, (now - Date.parse(ping.timestamp)) / 1000);
  const accuracy = ping.accuracy ?? 100;
  const battery = ping.battery;
  let score = 100;

  if (connection !== "live") score -= 35;
  if (ageSeconds > 120) score -= 40;
  else if (ageSeconds > 60) score -= 25;
  else if (ageSeconds > 20) score -= 10;

  if (accuracy > 80) score -= 25;
  else if (accuracy > 35) score -= 15;
  else if (accuracy > 15) score -= 7;

  if (battery !== null && battery < 10) score -= 25;
  else if (battery !== null && battery < 25) score -= 10;

  if (state === "alert") score -= 35;
  else if (state === "grace") score -= 20;
  else if (state === "caution") score -= 8;

  score = clamp(score);

  const level: CareConfidenceLevel =
    score >= 90 ? "excellent" : score >= 70 ? "good" : score >= 40 ? "attention" : "critical";
  const label =
    level === "excellent"
      ? "High confidence"
      : level === "good"
        ? "Reliable enough to act"
        : level === "attention"
          ? "Check tracking"
          : "Status uncertain";
  const summary =
    level === "excellent"
      ? "Location is recent, clear, and reporting normally."
      : level === "good"
        ? "The current location is usable, but one signal is less reliable."
        : level === "attention"
          ? "Open details to see what needs attention."
          : "Do not rely on the current status until the patient phone reconnects.";

  return {
    score,
    level,
    label,
    summary,
    details: [
      {
        label: "GPS",
        value: ping.accuracy ? (accuracy <= 12 ? "Strong" : accuracy <= 35 ? "Good" : "Approximate") : "Unknown",
        healthy: accuracy <= 35
      },
      {
        label: "Freshness",
        value: relativeUpdate(ping.timestamp),
        healthy: ageSeconds <= 60
      },
      {
        label: "Accuracy",
        value: ping.accuracy ? `±${Math.round(ping.accuracy)} m` : "Unknown",
        healthy: accuracy <= 35
      },
      {
        label: "Internet",
        value: connection === "live" ? "Stable" : "Reconnecting",
        healthy: connection === "live"
      },
      {
        label: "Battery",
        value: battery === null ? "Unavailable" : `${battery}%`,
        healthy: battery === null || battery >= 25
      },
      {
        label: "Safety state",
        value: STATE_CHIP_LABEL[state],
        healthy: state === "safe"
      }
    ]
  };
}
